Add "All Time" to scheduled statements date range; allow excluding clients with no matching statements

joshuadwire opened the pull request:
* Adds support for "All Time" option to scheduled statements
* Adds support for option to exclude clients with no matching invoices from scheduled statements

joshuadwire committed:
Support All Time option for statements and support excluding clients with no invoices matching the selected filters

turbo124 commented on app/Services/Scheduler/EmailStatementService.php:
```diff
@@ -92,6 +93,7 @@ private function calculateStartAndEndDates(): array
             EmailStatement::LAST_QUARTER => [now()->startOfDay()->subQuarterNoOverflow()->firstOfQuarter()->format('Y-m-d'), now()->startOfDay()->subQuarterNoOverflow()->lastOfQuarter()->format('Y-m-d')],
             EmailStatement::THIS_YEAR => [now()->startOfDay()->firstOfYear()->format('Y-m-d'), now()->startOfDay()->lastOfYear()->format('Y-m-d')],
             EmailStatement::LAST_YEAR => [now()->startOfDay()->subYearNoOverflow()->firstOfYear()->format('Y-m-d'), now()->startOfDay()->subYearNoOverflow()->lastOfYear()->format('Y-m-d')],
+            EmailStatement::ALL_TIME => [null, null],
```
Any reason why we can't use start_date = now() and end_date = epoch?

joshuadwire replied:
@turbo124 I think it would need to be the other way around, but also, the date is shown in the email and PDF templates. I added conditions to either hide the date range or show something else reasonable based on the invoices that were returned. If we set the dates here, then the email and PDF would show that the statement is from January 1, 1970 to now. Maybe that's OK though.

turbo124 replied:
I think the preferable approach would be to calculate the first and last invoice dates and then use this as the range. saves a lot of changes downstream also.

joshuadwire committed:
Switch "All Time" option for statements to pull the earliest invoice date.

joshuadwire commented:
@turbo124 I've updated the PR to calculate the first invoice date and use now() as the end date and reverted the other changes that are no longer needed. I think using now() as the end date is preferable to using the last invoice date, as that could confuse clients, thinking the statement isn't up to date.

turbo124 commented on app/Services/Scheduler/EmailStatementService.php:
```diff
@@ -92,6 +94,11 @@ private function calculateStartAndEndDates(): array
             EmailStatement::THIS_YEAR => [now()->startOfDay()->firstOfYear()->format('Y-m-d'), now()->startOfDay()->lastOfYear()->format('Y-m-d')],
             EmailStatement::LAST_YEAR => [now()->startOfDay()->subYearNoOverflow()->firstOfYear()->format('Y-m-d'), now()->startOfDay()->subYearNoOverflow()->lastOfYear()->format('Y-m-d')],
+            EmailStatement::ALL_TIME => [
+                $client->invoices()->selectRaw('MIN(invoices.date) as start_date')->pluck('start_date')->first()
```
The whereIn on $client->invoices() can be brutal for performance. Can we replace with

```
Invoice::withTrashed()->where('client_id', $client->id)->selectRaw('MIN(invoices.date) as start_date')->pluck('start_date')->first()
```

turbo124 commented:
Thanks, looks good, one tiny change needed.

joshuadwire committed:
Performance improvement

joshuadwire commented:
@turbo124 Ok, I've updated it. I'm curious, what's the difference between the two that creates a performance impact? As far as I'm aware, `->invoices()` should return roughly the same query, since I'm calling it as a function vs. using `->invoices` as a property. Is it related to some additional code you have related to multitenancy?

turbo124 commented:
@joshuadwire Sorry, this is my error, this is a hasMany() so you are correct.
